### ISSUE TITLE
Phase 46.2: MaskedPredictor — Masked Language & Image Modeling

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#906]
## Phase 46.2: MaskedPredictor

**Parent**: Phase 46 — Self-Supervised Learning & Contrastive Representations (Discussion #904)

### Overview
Implements masked prediction methods for self-supervised pre-training across language, vision, and multimodal domains. Learns representations by reconstructing masked portions of input data, following the paradigms established by BERT, MAE, and BEiT.

#### Masked Language Modeling (MLM)
- Random token masking with configurable mask ratio (default 15%)
- [MASK] token replacement strategy (80% mask, 10% random, 10% keep)
- Prediction head with tied/untied embedding weights
- Whole-word masking for subword tokenizers
- Span masking (SpanBERT-style contiguous masking)
- Next sentence prediction (NSP) auxiliary task

#### Masked Image Modeling (MIM)
- Patch-based masking with configurable patch size and mask ratio
- Pixel reconstruction targets (MAE-style L2 loss in pixel space)
- Discrete token targets (BEiT-style with visual tokenizer dVAE)
- Feature regression targets (data2vec-style with teacher features)
- Block-wise masking strategies (contiguous regions)
- High mask ratios (75%) for efficient pre-training

#### Masked Autoencoder Architecture
- Asymmetric encoder-decoder design
- Encoder processes only visible (unmasked) patches — computational efficiency
- Lightweight decoder reconstructs from latent + mask tokens
- Positional embeddings for spatial/temporal awareness
- Scalable to high-resolution inputs

#### Masking Strategy Optimization
- Random uniform masking (baseline)
- Block/grid structured masking
- Attention-guided masking (mask high-attention regions)
- Curriculum masking (easy → hard scheduling)
- Adversarial masking (mask most informative regions)

#### Multi-Modal Masked Prediction
- Joint text-image masking with cross-modal reconstruction
- Text-to-image and image-to-text prediction heads
- Aligned masking across modalities
- Shared vs. modality-specific decoders

### Interface
```python
class MaskedPredictor:
    def pretrain_masked(
        self,
        data: UnlabeledDataset,
        modality: str,  # "text" | "image" | "multimodal"
        mask_config: MaskConfig
    ) -> PretrainedModel:
        """Pre-train model via masked prediction."""

    def create_masked_batch(
        self,
        batch: Tensor,
        mask_ratio: float,
        strategy: str  # "random" | "block" | "attention" | "curriculum"
    ) -> Tuple[Tensor, Tensor, Tensor]:
        """Create masked input, targets, and mask positions."""

    def reconstruct(
        self,
        encoder_output: Tensor,
        mask_tokens: Tensor,
        decoder: nn.Module
    ) -> Tensor:
        """Reconstruct masked portions from encoder output."""
```

### Acceptance Criteria
- [ ] MLM achieves meaningful perplexity reduction on text corpus
- [ ] MIM reconstructs masked patches with low pixel-level error
- [ ] Asymmetric MAE encoder-decoder trains 3x+ faster than full encoder
- [ ] Block masking produces spatially coherent reconstructions
- [ ] Multi-modal masking improves cross-modal retrieval metrics
- [ ] Supports variable mask ratios (15%-90%) across modalities
- [ ] Unit tests with >90% coverage

### References
- Devlin et al. (2019) BERT — Masked Language Modeling
- He et al. (2022) MAE — Masked Autoencoders Are Scalable Vision Learners
- Bao et al. (2022) BEiT — BERT Pre-Training of Image Transformers
- Baevski et al. (2022) data2vec — General Framework for Self-Supervised Learning
